### PR TITLE
ACS-3525: API for manual triggering rules on a folder

### PR DESCRIPTION
**Pull request by krdabrowski**
https://alfresco.atlassian.net/browse/ACS-3525
This PR covers only basic E2E test. Further E2Es will be added within separate task.

### CODE REVIEW

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ExecuteRulesTests.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * Tests for POST /nodes/{nodeId}/rule-executions.
+ */
+@Test(groups = { TestGroup.RULES})
+public class ExecuteRulesTests extends RestTest
+{
+    private UserModel user;
```
Comment (mpichura) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ExecuteRulesTests.java:53]: Should we add `@Experimental` here as well?

File: remote-api/src/main/java/org/alfresco/rest/api/impl/rules/RulesImpl.java [hunk 2]
```diff
@@ -130,6 +132,26 @@ public void deleteRuleById(String folderNodeId, String ruleSetId, String ruleId)
...
+        final NodeRef folderNodeRef = validator.validateFolderNode(folderNodeId, false);
+        final Map<String, Serializable> parameterValues = new HashMap<>();
+        parameterValues.put(ExecuteAllRulesActionExecuter.PARAM_RUN_ALL_RULES_ON_CHILDREN, eachSubFolderIncluded);
+        parameterValues.put(ExecuteAllRulesActionExecuter.PARAM_EXECUTE_INHERITED_RULES, eachInheritedRuleExecuted);
+        final ActionImpl action = new ActionImpl(null, GUID.generate(), ExecuteAllRulesActionExecuter.NAME);
+        action.setNodeRef(folderNodeRef);
+        action.setParameterValues(parameterValues);
+
+        ActionAccessRestriction.setActionContext(action, ActionAccessRestriction.V1_ACTION_CONTEXT);
+        actionService.executeAction(action, folderNodeRef, true, false);
```
Comment (mpichura) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/rules/RulesImpl.java:147]: Could/should we check if at the moment any actions on given folder(s) are being executed (maybe this is somehow already happening in `actionService.executeAction` - don't know)?
In some cases action executions take quite some time and running them again could not be what we want.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ExecuteRulesTests.java (hunk 1 above)
Comment (krdabrowski) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ExecuteRulesTests.java:53]: Annotation `org.alfresco.service.Experimental` is not available from `alfresco-community-repo-restapi-test` module. None of our E2E tests is annotated as "experimental" so I think it's not necessary here as well.

**Review by mpichura — APPROVED** [review]

File: remote-api/src/main/java/org/alfresco/rest/api/model/rules/RuleExecution.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+import java.util.Objects;
+
+import org.alfresco.service.Experimental;
+
+@Experimental
+public class RuleExecution
+{
+    private boolean eachSubFolderIncluded;
+    private boolean eachInheritedRuleExecuted;
```
Comment (tpage-alfresco) [inline remote-api/src/main/java/org/alfresco/rest/api/model/rules/RuleExecution.java:36]: I thought that these were going to be `isEach...`?

Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/model/rules/RuleExecution.java:36]: Setters and getters simulates that this names looks like `isEach...`.

Comment (tpage-alfresco) [reply remote-api/src/main/java/org/alfresco/rest/api/model/rules/RuleExecution.java:36]: Ah - ok. So the setters and getters don't follow the Java convention. (Although I'm not sure having `isIs...` is a good getter name anyway :smile: ). I'm fine with this.

Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/model/rules/RuleExecution.java:36]: Luckily IDE has generated `getIs...` 😄

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ExecuteRulesTests.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        RestNodeModel node = restClient.authenticateUser(user).withCoreAPI().usingNode(file).getNode();
+        restClient.assertStatusCodeIs(HttpStatus.OK);
+        node.assertThat().field("aspectNames").notContains(AUDIO_ASPECT);
+
+        STEP("Execute rule");
+        restClient.authenticateUser(user).withCoreAPI().usingNode(ruleFolder).executeRules(createRuleExecutionRequest());
+        restClient.assertStatusCodeIs(HttpStatus.CREATED);
+
+        STEP("Check if file contains Audio aspect");
+        Utility.sleep(500, 10000, () -> {
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ExecuteRulesTests.java:88]: Is this sleep necessary?  I thought that rules would run immediately unless they're set to async.

Comment (krdabrowski) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ExecuteRulesTests.java:88]: Rules are added to pending list and afterwards executed by `RuleTransactionListener.beforeCommit()`. This seems to be immediately but I'm not 100% sure.

Comment (tpage-alfresco) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ExecuteRulesTests.java:88]: I think it's "immediate" - i.e. it happens in the same transaction.

Comment (krdabrowski) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ExecuteRulesTests.java:88]: Ok, I will remove the "sleep" method within PR adding rest of E2Es.

**Pull request merged**